src/Qir/Runtime/lib/Simulators/FullstateSimulator.cpp

```diff
@@ -408,7 +419,109 @@ namespace Quantum
+            {
+                std::ostream& outStream = *reinterpret_cast<std::ostream*>(location);
+
+                if (re != 0 || im != 0)
+                {
+                    outStream << "|" << std::bitset<8>(idx) << ">: " << re << "+" << im << "i" << std::endl;
+                }
+                return true;
+            };
+    }; // class CFullstateSimulator
+
+    void CFullstateSimulator::DumpMachineImpl(std::ostream& outStream)
+    {
+        outStream << "*********************" << std::endl;
+        this->GetStateTo((TDumpLocation)&outStream, dumpToLocationCallback);
+        outStream << "*********************" << std::endl;
+        outStream.flush();
+    }
+
+    void CFullstateSimulator::DumpRegisterImpl(std::ostream& outStream, const QirArray* qubits)
+    {
+        outStream << "*********************" << std::endl;
+        if(!this->GetRegisterTo((TDumpLocation)&outStream, dumpToLocationCallback, qubits))
+        {
+            outStream << "## Qubits were entangled with an external qubit. Cannot dump corresponding wave function. ##" << std::endl;
+        }
+        outStream << "*********************" << std::endl;
+        outStream.flush();
+    }
+
+    bool CFullstateSimulator::GetRegisterTo(TDumpLocation location, TDumpToLocationCallback callback, const QirArray* qubits)
+    {
+        // `qubits->buffer` points to the sequence of qubit ids each of type `void*` (logically `qubits->buffer` is of type `void**`).
+        // We need to pass to the native simulator the pointer to the sequence of `unsigned`.
+        std::vector<unsigned> unsQbIds(qubits->count);
+        size_t i = 0;
+        for (unsigned& qbId : unsQbIds)
+        {
+            qbId = (unsigned)(uintptr_t)(((void**)(qubits->buffer))[i]);
+            ++i;
+        }
+        static TDumpQubitsToLocationAPI dumpQubitsToLocation =
+            reinterpret_cast<TDumpQubitsToLocationAPI>(this->GetProc("DumpQubitsToLocation"));
+        return dumpQubitsToLocation(
+            this->simulatorId, (unsigned)(qubits->count), unsQbIds.data(), callback,
+            location);
+    }
+
+    void CFullstateSimulator::GetStateTo(TDumpLocation location, TDumpToLocationCallback callback)
+    {
+        static TDumpToLocationAPI dumpTo = reinterpret_cast<TDumpToLocationAPI>(this->GetProc("DumpToLocation"));
+        
+        dumpTo(this->simulatorId, callback, location);
+    }
+
+    std::ostream& CFullstateSimulator::GetOutStream(const void* location, std::ofstream& outFileStream)
+    {
+        // If the location is not nullptr and not empty string then dump to a file:
+        if((location != nullptr) &&
+            (((static_cast<const QirString *>(location))->str) != ""))
+        {
+            // Open the file for appending:
+            const std::string& filePath = (static_cast<const QirString *>(location))->str;
+
+            outFileStream.open(filePath, std::ofstream::out | std::ofstream::app);

```

Comment:
It looks like there's no exception handling for any I/O issues that could arise here; in general, diagnostic routines such as DumpMachine and DumpRegister should not cause a quantum program to fail, such that I would suggest catching I/O exceptions here and providing actionable error messages to the user at the console instead.

Reply:
Thanks. Good point.

The [`open()`](http://www.cplusplus.com/reference/fstream/ofstream/open/) throws exceptions if "member `exceptions` was set to throw". My code does not deal with "setting to throw" for this `std::ofstream` instance, that's why exception handling can be redundant. However, I see that this function gets the instance as a parameter, so "setting to throw" could have been done before this function. To make the function reliable I'm adding the exception handling.

Also, in case of `open()` failure, I'm adding the actionable complaint to `std::cerr` and dumping to `std::cout` (instead of calling `quantum__rt__fail()`).

Separate thanks for this feedback!